ntp-udp/src/lib.rs

```diff
@@ -1,219 +1,424 @@
+            mhdr.msg_namelen = 0;
+
+            // Safety:
+            // recvmsg ensures that the control buffer contains
+            // a set of valid control messages and that controllen is
+            // the length these take up in the buffer.
+            Ok((
+                sent_bytes,
+                unsafe { ControlMessageIterator::new(mhdr) },
+                sockaddr_storage_to_socket_addr(&addr),
+            ))
         }
-    })
-}
 
-/// The space used to store a control message that contains a value of type T
-const fn control_message_space<T>() -> usize {
-    (unsafe { libc::CMSG_SPACE((std::mem::size_of::<T>()) as _) }) as usize
-}
+        struct ControlMessageIterator<'a> {
+            mhdr: libc::msghdr,
+            current_msg: *const libc::cmsghdr,
+            phantom: PhantomData<&'a [u8]>,
+        }
 
-pub(crate) fn zeroed_sockaddr_storage() -> libc::sockaddr_storage {
-    // a zeroed-out sockaddr storage is semantically valid, because a ss_family with value 0 is
-    // libc::AF_UNSPEC. Hence the rest of the data does not come with any constraints
-    unsafe { std::mem::MaybeUninit::zeroed().assume_init() }
-}
+        impl<'a> ControlMessageIterator<'a> {
+            // Safety assumptions:
+            // mhdr has a control and controllen field
+            // that together describe a memory region
+            // with lifetime 'a containing valid control
+            // messages
+            unsafe fn new(mhdr: libc::msghdr) -> Self {
+                // Safety:
+                // mhdr's control and controllen fields are valid and point
+                // to control messages.
+                let current_msg = unsafe { libc::CMSG_FIRSTHDR(&mhdr) };
+                Self {
+                    mhdr,
+                    current_msg,
+                    phantom: PhantomData,
+                }
+            }
+        }
 
-#[derive(Debug, Clone, Copy, Default)]
-pub(crate) struct TimestampingConfig {
-    rx_software: bool,
-    tx_software: bool,
-}
+        pub(crate) enum ControlMessage {
+            Timestamping(libc::timespec),
+            ReceiveError(libc::sock_extended_err),
+            Other(libc::cmsghdr),
+        }
 
-#[repr(C)]
-#[allow(non_camel_case_types)]
-#[derive(Default)]
-struct ethtool_ts_info {
-    cmd: u32,
-    so_timestamping: u32,
-    phc_index: u32,
-    tx_types: u32,
-    tx_reserved: [u32; 3],
-    rx_filters: u32,
-    rx_reserved: [u32; 3],
-}
+        impl<'a> Iterator for ControlMessageIterator<'a> {
+            type Item = ControlMessage;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                // Safety:
+                // CMSG_FIRSTHDR and CMSG_NXTHDR only return valid pointers or NULL when given valid input
+                let current_msg = unsafe { self.current_msg.as_ref() };
+                if let Some(current_msg) = current_msg {
+                    // Safety:
+                    // New ensure mhdr is valid
```

Comment:
It's not clear to me what is meant by "New ensure" at different levels.

Reply:
I agree that the previous explanation was unclear, have rewritten the entire safety argument around a set of two invariants holding for the `ControlMessageIterator` struct.